From 0bf23a877046f337f4ac334a0ab24c3e492d714a Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=E9=98=BF=E8=8F=8C=E2=80=A2=E6=9C=AA=E9=9C=9C?=
 <799620521@qq.com>
Date: Thu, 2 Mar 2023 11:59:10 +0800
Subject: [PATCH] mm: uksm: Don't hard code MAX time ratio scale

---
 mm/uksm.c | 64 +++++++++++++++++++++++++++++++++++++++++++++++++++----
 1 file changed, 60 insertions(+), 4 deletions(-)

diff --git a/mm/uksm.c b/mm/uksm.c
index a2cb54470d878..9a9551c62a80d 100644
--- a/mm/uksm.c
+++ b/mm/uksm.c
@@ -536,10 +536,66 @@ struct uksm_cpu_preset_s {
 };
 
 struct uksm_cpu_preset_s uksm_cpu_preset[4] = {
-	{ {20, 40, -2500, -10000}, {1000, 500, 200, 50}, 95},
-	{ {20, 30, -2500, -10000}, {1000, 500, 400, 100}, 50},
-	{ {10, 20, -5000, -10000}, {1500, 1000, 1000, 250}, 20},
-	{ {10, 20, 40, 75}, {2000, 1000, 1000, 1000}, 1},
+	{
+		{
+			20,
+			40,
+			-TIME_RATIO_SCALE / 4,
+			-TIME_RATIO_SCALE
+		},
+		{
+			1000,
+			500,
+			200,
+			50
+		},
+		95
+	},
+	{
+		{
+			20,
+			30,
+			-TIME_RATIO_SCALE / 4,
+			-TIME_RATIO_SCALE
+		},
+		{
+			1000,
+			500,
+			400,
+			100
+		},
+		50
+	},
+	{
+		{
+			10,
+			20,
+			-TIME_RATIO_SCALE / 2,
+			-TIME_RATIO_SCALE
+		},
+		{
+			1500,
+			1000,
+			1000,
+			250
+		},
+		20
+	},
+	{
+		{
+			10,
+			20,
+			40,
+			75
+		},
+		{
+			2000,
+			1000,
+			1000,
+			1000
+		},
+		1
+	},
 };
 
 /* The default value for uksm_ema_page_time if it's not initialized */
